Framework/PythonInterface/plugins/algorithms/HB3AFindPeaks.py

```diff
@@ -0,0 +1,164 @@
+                    issues[param] = "Lattice parameter must be greater than 0"
+                if param == "LatticeAlpha" or param == "LatticeBeta" or param == "LatticeGamma":
+                    if prop.value < 5.0 or prop.value > 175.0:
+                        issues[param] = "Lattice parameter must have an angle between 5 and 175."
+                set_count += 1
+
+        if 0 <= set_count < 6 and not self.getProperty("UseLattice").isDefault:
+            for param in lattice_params:
+                issues[param] = "Must set all or none of the optional lattice parameters."
+
+        if self.getProperty("InputWorkspace").value.getNumDims() != 3:
+            issues["InputWorkspace"] = "Workspace has the wrong number of dimensions"
+
+        return issues
+
+    def PyExec(self):
+        input_ws = self.getProperty("InputWorkspace").value
+        cell_type = self.getProperty("CellType").value
+        centering = self.getProperty("Centering").value
+
+        npeaks = self.getProperty("MaxPeaks").value
+        dist_thresh = self.getProperty("PeakDistanceThreshold").value
+        density_thresh = self.getProperty("DensityThresholdFactor").value
+
+        lattice = self.getProperty("UseLattice").value
+        if lattice:
+            a = self.getProperty("LatticeA").value
+            b = self.getProperty("LatticeB").value
+            c = self.getProperty("LatticeC").value
+            alpha = self.getProperty("LatticeAlpha").value
+            beta = self.getProperty("LatticeBeta").value
+            gamma = self.getProperty("LatticeGamma").value
+
+        peak_ws = self.getProperty("OutputWorkspace")
+
+        # Initially set the back-up wavelength to use. This is overwritten if found in sample logs.
+        wavelength = self.getProperty("Wavelength").value
+
+        if input_ws.getNumExperimentInfo() == 0:
+            # Warn if we could extract a wavelength from the workspace
+            raise RuntimeWarning("No experiment info was found in input '{}'".format(input_ws.getName()))
+        else:
+            exp_info = input_ws.getExperimentInfo(0)
+            if exp_info.run().hasProperty("wavelength"):
+                wavelength = exp_info.run().getProperty("wavelength").value
+
+        peak_ws = FindPeaksMD(InputWorkspace=input_ws,
+                              PeakDistanceThreshold=dist_thresh, DensityThresholdFactor=density_thresh,
+                              CalculateGoniometerForCW=True,
+                              Wavelength=wavelength,
+                              FlipX=True, InnerGoniometer=True,
+                              MaxPeaks=npeaks)
+
+        IndexPeaks(PeaksWorkspace=peak_ws)
+
+        if lattice:
+            FindUBUsingLatticeParameters(PeaksWorkspace=peak_ws, a=a, b=b, c=c, alpha=alpha, beta=beta, gamma=gamma)
+        else:
+            # Using IndexedPeaks seems to work in some cases where FindUBUsingFFT fails sometimes..?
+            FindUBUsingIndexedPeaks(PeaksWorkspace=peak_ws)
```

Comment:
This and `IndexPeaks` only works if we already have the UB matrix, in all the test data we have been given that is the case but we can't depend on it. Also if there is a UB be can't assume that it is correct, so I think we need to use FindUBUsingFFT in those cases.